### PR TITLE
Implement `internal_bit`

### LINKED ISSUE #30: Implement `internal_bit`
We have to only implement `ceil_pow2`, which should equal to ~~`|n| 32 - n.leading_zeros()`~~ `|n| 32 - n.saturating_sub(1).leading_zeros()`.

### PR DESCRIPTION
**Pull request by qryxip**
Closes #30.

### CODE REVIEW

**Comment by qryxip** [comment]
Seeing the usage of `ceil_pow2` in the original ACL, I think the signature should be `fn(u32) -> u32`.

**Comment by TonalidadeHidrica** [comment]
LGTM.  Since three noodles depends on this modules, I think this PR should be merged ASAP.

File: src/internal_bit.rs [hunk 1]
```diff
@@ -1 +1,30 @@
+// Skipped:
+//
+// - `bsf` = `__builtin_ctz`: should equivalent to `{integer}::trailing_zeros`
```
Comment (koba-e964) [inline src/internal_bit.rs:3]: Stylistic issue:
```suggestion
// - `bsf` = `__builtin_ctz`: is equivalent to `{integer}::trailing_zeros`
```

**Review by koba-e964 — APPROVED** [review]
LGTM with minor issues

File: src/internal_bit.rs (hunk 1 above)
Comment (qryxip) [reply src/internal_bit.rs:3]: thank you

**Comment by qryxip** [comment]
```console
$ cargo +nightly bench --bench bench
   Compiling a v0.1.0 (/home/ryo/src/local/hage/a)
    Finished bench [optimized] target(s) in 0.50s
     Running target/release/deps/bench-13a9f15c9072cc25

running 2 tests
test naive              ... bench:   2,115,928 ns/iter (+/- 117,993)
test with_leading_zeros ... bench:     104,699 ns/iter (+/- 8,538)

test result: ok. 0 passed; 0 failed; 0 ignored; 2 measured; 0 filtered out
```

**Comment by TonalidadeHidrica** [comment]
@qryxip Just for curiosity, could you tell me which two function you compared?  The original one (while loop + comparison) and the Rust implementation (count leading zeros)?

**Comment by qryxip** [comment]
This. (I'm not used to `bench`)

```rust
#![feature(test)]

extern crate test;

use test::Bencher;

#[bench]
fn with_leading_zeros(bencher: &mut Bencher) {
    let values = values();
    bencher.iter(move || {
        values
            .iter()
            .map(|v| 32 - v.saturating_sub(1).leading_zeros())
            .fold(0, u32::wrapping_add)
    });
}

#[bench]
fn naive(bencher: &mut Bencher) {
    let values = values();
    bencher.iter(move || {
        values
            .iter()
            .map(|&v| {
                let mut x = 0;
                while (1 << x) < v {
                    x += 1;
                }
                x
            })
            .fold(0, u32::wrapping_add)
    });
}

fn values() -> Vec<u32> {
    std::fs::read_to_string("./values.txt")
        .unwrap()
        .split_ascii_whitespace()
        .map(|s| s.parse().unwrap())
        .collect()
}
```

```python
import random
s = ''
for _ in range(100000):
    s += str(random.randint(0, 1 << 30))
    s += '\n'
with open('./values.txt', 'w') as file:
    file.write(s)
```

**Comment by qryxip** [comment]
```console
$ cargo +nightly bench --bench bench
   Compiling a v0.1.0 (/home/ryo/src/local/hage/a)
    Finished bench [optimized] target(s) in 0.45s
     Running target/release/deps/bench-13a9f15c9072cc25

running 2 tests
test naive              ... bench:   2,144,545 ns/iter (+/- 99,698)
test with_leading_zeros ... bench:     144,275 ns/iter (+/- 11,680)

test result: ok. 0 passed; 0 failed; 0 ignored; 2 measured; 0 filtered out
```

```rust
#[bench]
fn with_leading_zeros(bencher: &mut Bencher) {
    #[inline(never)]
    fn run(v: u32) -> u32 {
        32 - v.saturating_sub(1).leading_zeros()
    }

    let values = values();
    bencher.iter(move || values.iter().copied().map(run).fold(0, u32::wrapping_add));
}

#[bench]
fn naive(bencher: &mut Bencher) {
    #[inline(never)]
    fn run(v: u32) -> u32 {
        let mut x = 0;
        while (1 << x) < v {
            x += 1;
        }
        x
    }

    let values = values();
    bencher.iter(move || values.iter().copied().map(run).fold(0, u32::wrapping_add));
}
```

**Comment by TonalidadeHidrica** [comment]
Can somebody merge this so that others can implement (lazy)segtree ?

**Pull request merged**